### PR TITLE
Special scrub handling of hinfo_key errors

### PR DESCRIPTION
**Pull request by dzafman**

### CODE REVIEW

File: src/include/rados/rados_types.hpp [hunk 1]
```diff
@@ -55,21 +55,31 @@ struct err_t {
     SHARD_MISSING        = 1 << 1,
     SHARD_STAT_ERR       = 1 << 2,
     SHARD_READ_ERR       = 1 << 3,
-    DATA_DIGEST_MISMATCH_OI = 1 << 9,
-    OMAP_DIGEST_MISMATCH_OI = 1 << 10,
-    SIZE_MISMATCH_OI        = 1 << 11,
+    DATA_DIGEST_MISMATCH_OI = 1 << 9,   // Old
```
Comment (dzafman) [inline src/include/rados/rados_types.hpp]: Should I keep these old names for compilation of older external code?

File: src/tools/rados/rados.cc [hunk 2]
```diff
@@ -1369,35 +1375,67 @@ static void dump_shard(const shard_info_t& shard,
...
+       || inc.has_hinfo_inconsistency()) &&
+       !shard.has_hinfo_missing()) {
+    map<std::string, ceph::bufferlist>::iterator k = (const_cast<shard_info_t&>(shard)).attrs.find(ECUtil::get_hinfo_key());
+    assert(k != shard.attrs.end()); // Can't be missing
+    if (!shard.has_hinfo_corrupted()) {
+      ECUtil::HashInfo hi;
+      bufferlist bl;
+      bufferlist::iterator bliter = k->second.begin();
+      decode(hi, bliter);  // Can't be corrupted
+      f.dump_stream("hashinfo") << hi;
```
Comment (dzafman) [inline src/tools/rados/rados.cc]: Instead of using operator<< should we output JSON for the HashInfo.

File: src/tools/rados/rados.cc [hunk 3]
```diff
@@ -1369,35 +1375,67 @@ static void dump_shard(const shard_info_t& shard,
...
-     || inc.union_shards.has_oi_attr_missing()
-     || inc.union_shards.has_oi_attr_corrupted()
-     || inc.union_shards.has_ss_attr_missing()
-     || inc.union_shards.has_ss_attr_corrupted()) {
+    if (!shard.has_snapset_corrupted()) {
+      SnapSet ss;
+      bufferlist bl;
+      bufferlist::iterator bliter = k->second.begin();
+      decode(ss, bliter);  // Can't be corrupted
+      f.dump_stream("snapset") << ss;
```
Comment (dzafman) [inline src/tools/rados/rados.cc]: Update this too from operator<< to JSON for SnapSet?

**Comment by dzafman** [comment]
Want to also rename ss_attr_corrupted in the list-inconsistent-snapset and make sure that non-primary corruption is also reflected their.

**Comment by dzafman** [comment]
It would be a much more involved change to handle all shards when dealing with snapshot analysis.  I filed tracker http://tracker.ceph.com/issues/23428 to at least indicate which authoritative shard was used to determine snapshot consistency.

**Comment by dzafman** [comment]
> Make sure that non-authoritative corruption is also reflected in list-inconsistent-snapset.

This is a lot of work which I don't think is a priority right now.

File: src/tools/rados/rados.cc (hunk 3 above)
Comment (jdurgin) [reply src/tools/rados/rados.cc]: yeah, sounds like a good idea

File: src/tools/rados/rados.cc (hunk 2 above)
Comment (jdurgin) [reply src/tools/rados/rados.cc]: sure

**Comment by jdurgin** [comment]
This looks good, I agree json for the snapset and hashinfo would be nice as well

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-03-28_14:42:54-rados-wip-kefu-testing-2018-03-28-1948-distro-basic-smithi/

**Comment by tchaikov** [comment]
it's good to merge if we want to improve the json representation of snapset and hashinfo in another PR.

**Comment by dzafman** [comment]
@jdurgin I rebased this branch.  Changed the system xattrs to output json format.  Added a fix for tracker 23428.

Concerns for object_info/hashinfo/snapset:

- [ ] It is easier to visually compare the differences in the single line format.  Just use that format?
- [ ] Keep the strings using old name (e.g. "object_info") for compatibility and add "*_json" for the json?
- [ ] I noticed that the single line format doesn't show all fields
- [ ] In the json output I'd like to show the flags as an array of strings instead an integer

**Comment by dzafman** [comment]
retest this please - run-tox-mgr-dashboard (Failed)

**Comment by jdurgin** [comment]
making the snapset/hash_info/object_info json means it's machine-readable, instead of having to parse it again. this is worth it imo, since the single-line format isn't legible to folks who don't know the code already.

Using flag names instead of an int is a good idea. I'm fine keeping the same field names, just add something to PendingReleaseNotes about it.

So basically, this looks good to me, it just needs a release note.

**Comment by dzafman** [comment]
@jdurgin I had to rebase because of conflicts with the PendingReleaseNotes

**Comment by dzafman** [comment]
retest this please

File: PendingReleaseNotes [hunk 4]
```diff
@@ -126,3 +126,42 @@ method. See http://docs.ceph.com/docs/luminous/mgr/restful for details.
...
+    is corrupt then "object_info" is a string containing the value.
+
+* *rados list-inconsistent-snapset format changes:*
+
+  * Various error strings have been improved.  For example, the "ss_attr" in
+    errors which stands for snapset info is now "snapset" (e.g. ss_attr_missing is
+    now snapset_missing).  The error snapset_mismatch has been renamed to snapset_error
+    to better reflect what it means.
+
+  * The head snapset information is output in json format as "snapset."  This means that
```
Comment (tchaikov) [inline PendingReleaseNotes]: "snapset".

**Pull request merged**